### DISCUSSION TITLE
builder.Services.AddDbContext<MyContext>(ServiceLifetime.Transient)?

### DISCUSSION THREAD

**Discussion by michaelcsikos** [MarimerLLC/csla#3209] [Questions] [answered]
In a Blazor solution, sometimes I get the following error:

`A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext.`

Is the solution `ServiceLifetime.Transient` or are there side effects to this approach?

**Comment by rockfordlhotka** [comment]
Transient lifetime means that you get a new instance of the service each time you ask for one.

Scoped lifetime means that you get a new instance of the service if there isn't already one in the current scope.

Singleton lifetime means that you get a new instance of the service only once during the lifetime of the app.

The local data portal creates a new scope when the code hits the logical server-side code, and disposes that scope before returning to the logical client. All server-side code executing during each data portal call runs in that scope.

Generally speaking, a remote data portal call works the same way, because aspnetcore creates a scope, the server-side code runs in that scope, and the scope is disposed at the end of the call.

**Comment by jhuerta3** [comment]
Hello @michaelcsikos ,

Im getting a error similar to this topic. I working on a Blazor control with pagination. May I ask how you resolved the issue? My error is below. 

A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.

**Reply by michaelcsikos** [reply]
I have been using `ServiceLifetime.Transient`. In `Program.cs`:

```cs
builder.Services.AddDbContext<EfDbContext>(options => options.UseSqlServer(builder.Environment.IsProduction()
                                                                           ? PRODUCTION_CONNECTION_STRING
                                                                           : DEVELOPMENT_CONNECTION_STRING),
                                           ServiceLifetime.Transient);
```

Each of my `Dal` classes call `DbContext.SaveChanges()` for their own changes so I'm quite happy for each one to get a new instance. It's much better than getting this horrible exception.

**Reply by jhuerta3** [reply, marked as answer]
Thank you. 
I just realized I was using the following on my Program.cs 
builder.Services.AddSingleton<ASPNetContext>(s => new ASPNetContext(aspnetdb_connectionString));
so I changed it to 
builder.Services.AddTransient<ASPNetContext>(s => new ASPNetContext(aspnetdb_connectionString));
Thank you  @michaelcsikos
